### PR TITLE
Fix/formatting

### PR DESCRIPTION
**Pull request by sum2000**
I decided to pass the format object to the new methods instead of interface since that is easier for the implementer. Let me know if you think the interface makes more sense and why.

### CODE REVIEW

File: eris_test.go [hunk 1]
```diff
@@ -35,17 +35,17 @@ func TestErrorWrapping(t *testing.T) {
 		"standard error wrapping with global root cause (eris.NewGlobal)": {
 			cause:  globalErr,
 			input:  []string{"additional context", "even more context"},
-			output: "global error: additional context: even more context",
+			output: "global error:additional context:even more context",
```
Comment (morningvera) [inline eris_test.go]: this doesn't look very good. any way we can still have a space when the stack is not included?

File: eris_test.go [hunk 2]
```diff
@@ -61,9 +61,9 @@ func TestErrorWrapping(t *testing.T) {
 		t.Run(desc, func(t *testing.T) {
 			err := setupTestCase(false, tc.cause, tc.input)
 			if err != nil && tc.cause == nil {
-				t.Errorf("%v: wrapping nil errors should return nil but got { %v }", desc, err)
+				t.Errorf("%v:wrapping nil errors should return nil but got { %v }", desc, err)
```
Comment (morningvera) [inline eris_test.go]: guessing this is a find/replace issue. add spaces back for t.Errorf calls.

File: eris_test.go [hunk 3]
```diff
@@ -228,25 +228,25 @@ func TestErrorFormatting(t *testing.T) {
...
 			} else if err != nil && tc.output != err.Error() {
-				t.Errorf("%v: expected { %v } got { %v }", desc, tc.output, err)
+				t.Errorf("%v:expected { %v } got { %v }", desc, tc.output, err)
 			}
 
-			// todo: automate stack trace verification
+			// todo:automate stack trace verification
```
Comment (morningvera) [inline eris_test.go]: same with this.

File: examples_test.go [hunk 4]
```diff
@@ -19,9 +19,7 @@ func ExampleUnpackedError_ToJSON_external() {
 
 	// unpack and print the error
 	err := readFile("example.json")
-	uerr := eris.Unpack(err)
-	format := eris.NewDefaultFormat(false) // false: omit stack trace
-	u, _ := json.Marshal(uerr.ToJSON(format))
+	u, _ := json.Marshal(eris.ToJSON(err, false)) // false: omit stack trace
```
Comment (morningvera) [inline examples_test.go]: Fix all example links. You need to change their names so they're correctly linked to methods in godoc. e.g. ExampleUnpackedError_ToJSON_external -> ExampleToJSON_external

File: examples_test.go [hunk 5]
```diff
@@ -49,26 +47,24 @@ func ExampleUnpackedError_ToJSON_global() {
...
+	// 	"root":{
+	// 		"message":"unexpected EOF",
+	// 		"stack":[
+	// 			"main.readFile:.../example/main.go:6",
+	// 			"main.parseFile:.../example/main.go:12",
+	// 			"main.main:.../example/main.go:20",
 	// 		]
 	// 	},
-	// 	"wrap": [
+	// 	"wrap":[
```
Comment (morningvera) [inline examples_test.go]: more find/replace issues?

File: format.go [hunk 6]
```diff
@@ -6,24 +6,24 @@ import (
...
+	StackElemSep string // Separator between elements of each stack frame.
+	ErrorSep     string // Separator between each error in the chain.
 }
 
 // NewDefaultFormat conveniently returns a basic format for the default string formatter.
 func NewDefaultFormat(withTrace bool) Format {
 	stringFmt := Format{
 		WithTrace: withTrace,
-		Sep:       ": ",
+		ErrorSep:  ":",
```
Comment (morningvera) [inline format.go]: Maybe we should change the sep to `: ` if withTrace is false

File: format.go [hunk 7]
```diff
@@ -55,19 +55,28 @@ func Unpack(err error) UnpackedError {
 	return upErr
 }
 
-// ToString returns a default formatted string for a given eris error.
-func (upErr *UnpackedError) ToString(format Format) string {
+// ToFormattedString returns a custom formatted string for a given eris error.
```
Comment (morningvera) [inline format.go]: Consider changing this to ToCustomString (same with JSON)

File: format.go [hunk 8]
```diff
@@ -55,19 +55,28 @@ func Unpack(err error) UnpackedError {
...
+//
+// To declare custom format, the Format object has to be passed as an argument.
+// An error without trace will be formatted as <Root error msg>[Format.ErrorSep]<Wrap error msg>
+// An error with trace will be formatted as following
+// <Root error msg>[Format.MsgStackSep]
+// [Format.PreStackSep]<Method1>[Format.StackElemSep]<File1>[Format.StackElemSep]<Line1>[Format.ErrorSep]
```
Comment (morningvera) [inline format.go]: should we also show method 2 under root error since it'll be added to the stack?

File: format.go [hunk 9]
```diff
@@ -79,8 +88,49 @@ func (upErr *UnpackedError) ToString(format Format) string {
...
 }
 
-// ToJSON returns a JSON formatted map for a given eris error.
-func (upErr *UnpackedError) ToJSON(format Format) map[string]interface{} {
+// ToString returns a default formatted string for a given eris error.
+//
+// An error without trace will be formatted as <Root error msg>:<Wrap error msg> for the default format.
+// An error with trace will be formatted as following for the default format.
+// <Root error msg>\n
+// \t<Method1>:<File1>:<Line1>\n
```
Comment (morningvera) [inline format.go]: change ":" to something like "[Format. StackElemSep]"

File: format.go [hunk 10]
```diff
@@ -79,8 +88,49 @@ func (upErr *UnpackedError) ToString(format Format) string {
...
+func ToString(err error, withTrace bool) string {
+	return ToFormattedString(err, NewDefaultFormat(withTrace))
+}
+
+// ToFormattedJSON returns a JSON formatted map for a given eris error.
+//
+// To declare custom format, the Format object has to be passed as an argument.
+// An error without trace will be formatted as following
+// 	{
+// 	"root": {
```
Comment (morningvera) [inline format.go]: fix indentation

File: format.go [hunk 11]
```diff
@@ -79,8 +88,49 @@ func (upErr *UnpackedError) ToString(format Format) string {
...
+// 		{
+// 			"message": "Wrap error msg'",
+// 		}
+// 	]
+// }
+// An error with trace will be formatted as following
+// 	{
+// 	"root": {
+// 		"message": "Root error msg",
+//		"stack": [ "<Method1>[Format.StackElemSep]<File1>[Format.StackElemSep]<Line1>",
```
Comment (morningvera) [inline format.go]: fix indentation here too

File: format.go [hunk 12]
```diff
@@ -102,6 +152,38 @@ func (upErr *UnpackedError) ToJSON(format Format) map[string]interface{} {
 	return jsonMap
 }
 
+// ToJSON returns a JSON formatted map for a given eris error.
+//
+// An error without trace will be formatted as following
+// 	{
+// 	"root": {
+// 		"message": "Root error msg",
```
Comment (morningvera) [inline format.go]: fix indentation

File: format.go [hunk 13]
```diff
@@ -102,6 +152,38 @@ func (upErr *UnpackedError) ToJSON(format Format) map[string]interface{} {
...
+// 			"message": "Wrap error msg'",
+// 		}
+// 	]
+// }
+// An error with trace will be formatted as following
+// 	{
+// 	"root": {
+// 		"message": "Root error msg",
+//		"stack": [ "<Method1>:<File1>:<Line1>",
+//				   "<Method2>:<File2>:<Line2>"
```
Comment (morningvera) [inline format.go]: same here

File: eris_test.go (hunk 1 above)
Comment (sum2000) [reply eris_test.go]: hmm, have to pass different formatter when trace is false. I will try

File: format.go (hunk 9 above)
Comment (sum2000) [reply format.go]: for default format, I decided to go with the resolved values rather than the object for clarity, is it not clear?

Comment (morningvera) [reply format.go]: oh, didn't realize this is for default. this is fine as is

**Pull request merged**